source/simulator/helper_functions.cc

```diff
@@ -2403,6 +2404,235 @@ namespace aspect
+                                                fe_values[introspection.extractors.compositional_fields[0]].value(j,q) *
+                                                fe_values.JxW(q);
+                    }
+                }
+            }
+
+
+          // SURFACE TERM
+          // TODO: the solution at the surface does not agree with the benchmark.
+          // this could be because of the calculation here or the assembly of the RHS in assemblers/adjoint.cc
+
+          // see if the cell is at the *top* boundary and if so start surface term calculation
+          unsigned int top_face_idx = numbers::invalid_unsigned_int;
+          for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+            if (cell->at_boundary(f) && geometry_model->depth (cell->face(f)->center()) < cell->face(f)->minimum_vertex_distance()/3)
+              {
+                top_face_idx = f;
+                break;
+              }
+
+          if (top_face_idx != numbers::invalid_unsigned_int)
+            {
+              fe_face_values.reinit (cell,top_face_idx);
+
+              // Evaluate the material properties and the solution at the cell boundary
+              MaterialModel::MaterialModelInputs<dim> in_face(fe_face_values, cell, introspection, solution);
+              MaterialModel::MaterialModelOutputs<dim> out_face(fe_face_values.n_quadrature_points, introspection.n_compositional_fields);
+              material_model->evaluate(in_face, out_face);
+
+              fe_face_values[introspection.extractors.temperature].get_function_values(topo_vector, topo_values);
+
+              for (unsigned int q=0; q<n_q_face_points; ++q)
+                {
+                  Point<dim> location = fe_face_values.quadrature_point(q);
+
+                  // get the relevant parameters
+                  const double density   = out_face.densities[q];
+                  const double eta       = out_face.viscosities[q];
+
+                  const SymmetricTensor<2,dim> strain_rate = in_face.strain_rate[q] - 1./3 * trace(in_face.strain_rate[q]) * unit_symmetric_tensor<dim>();
+                  const Tensor<1,dim> gravity = gravity_model->gravity_vector(location);
+                  const Tensor<1,dim> n_hat = - gravity/gravity.norm();
+
+                  for (unsigned int k=0; k<dofs_per_cell; ++k)
+                    {
+                      phi_rho[k] = fe_values[introspection.extractors.compositional_fields[0]].value(k,q);
+                      phi_eta[k] = fe_values[introspection.extractors.compositional_fields[1]].value(k,q);
+                    }
+
+                  for (unsigned int i = 0; i<dofs_per_cell; ++i)
+                    {
+                      // calculate the density term
+                      local_rhs(i) += - topo_values[q] * topo_values[q] / (density - density_above) * phi_eta[i] * fe_face_values.JxW(q);
```

Comment:
precompute JxW outside of this loop for here and use it here and in the line below

Reply:
What happens if `density == density_above`? Or does this never happen because density_above is either 0 (air) or 1000 (water)?

Reply:
For any physically reasonable problem, the two densities should never be the same. Should I add an assert that checks that? We currently use the same equations in the DT postprocessor and it doesn't check that there either.

Reply:
This is now being checked in the DT postprocessor. Pull request #4241 by Cedric.